ZEP-2225: Bluetooth: Ability to unregister GATT services

Vudentz opened the pull request:
In order to properly support unregistering GATT services at runtime we must support service changed characteristic thus this set first introduces support for it to later add bt_gatt_svc which greatly simplify the procedure to remove services and as a bonus reduces the size of bt_gatt_attr.

Jira: ZEP-2225

Vudentz committed:
Bluetooth: Remove CONFIG_BLUETOOTH_GATT_DYNAMIC_DB

Removes CONFIG_BLUETOOTH_GATT_DYNAMIC_DB in preparation to the
introduction of bt_gatt_unregister.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Kconfig: Add CONFIG_BLUETOOTH_DEVICE_APPEARANCE

This makes applications able to select the value used for the
GAP appearance attribute.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add GAP service by default

GAP is mandatory service and now that the db can only be build
dynamically there is no reason to keep the applications registering it.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add GATT service by default

GATT is mandatory service and now that the db can only be build
dynamically there is no reason to keep the applications registering it.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Indicate Service Changed when a service is added

If the database is changed indicate the range changed to all peers
subscribed.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add bt_gatt_register_service

This adds bt_gatt_register_service using bt_gatt_service which contains
the attribute array that is then added to the database saving a pointer
in each and every attribute declared.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add bt_gatt_unregister_service

With the introduction of Service Changed support it is now possible to
unregister services at runtime.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: Shell: Add gatt-unregister-service command

gatt-unregister-service can be used to remove the test service at
runtime causing service changed to be indicated:

00:1b:dc:07:31:88 (public)> gatt-unregister-service
[bt] [DBG] gatt_indicate: (0x0011e100) conn 0x0011d880 handle 0x0008
[bt] [DBG] bt_att_req_send: (0x0011e100) conn 0x0011d880 req 0x0011db00
[bt] [DBG] att_send_req: (0x0011e100) req 0x0011db00
Unregistering test vendor service

< ACL Data TX: Handle 3585 flags 0x00 dlen 11
      ATT: Handle Value Indication (0x1d) len 6
        Handle: 0x0008
          Data: 0a001000
> ACL Data RX: Handle 3585 flags 0x02 dlen 5
      ATT: Handle Value Confirmation (0x1e) len 0

Jira: ZEP-2225
Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on include/misc/slist.h:
```diff
@@ -415,7 +415,7 @@ static inline void sys_slist_remove(sys_slist_t *list,
  * @param list A pointer on the list to affect
  * @param node A pointer on the node to remove from the list
  */
```
You'll need to document the new return parameter and its meaning here.

jhedberg replied:
I fixed this myself (I had apparently permissions to push --force to your PR branch)

jhedberg commented on samples/bluetooth/peripheral/prj.conf:
```diff
@@ -4,6 +4,6 @@ CONFIG_BLUETOOTH_SMP=y
-CONFIG_BLUETOOTH_GATT_DYNAMIC_DB=y
 CONFIG_BLUETOOTH_PRIVACY=y
 CONFIG_BLUETOOTH_DEVICE_NAME="Test peripheral"
+CONFIG_BLUETOOTH_DEVICE_APPEARANCE=0x0341
```
This doesn't actually work. You've declared the Kconfig as 'int' instead of 'hex' and so the above results in the following in .config:

CONFIG_BLUETOOTH_DEVICE_APPEARANCE=0

jhedberg replied:
Since the assigned numbers use decimal integers I'd say your Kconfig definition is ok, and the hex versions should simply be converted to decimal when moved from the code to the prj.conf.

jhedberg replied:
I went ahead and fixed this one as well using push --force to your branch

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -280,7 +286,29 @@ struct bt_gatt_cpf {
-int bt_gatt_register(struct bt_gatt_attr *attrs, size_t count);
+#define bt_gatt_register(_attrs, _count)				\
+	bt_gatt_register_service((&(struct bt_gatt_svc) { .attrs = _attrs, \
+				.attr_count = _count, }))
```
How can this work? The anonymous variable you're creating here will be an automatic one, i.e. become invalid as soon as the function that calls bt_gatt_register() returns, e.g. if you call bt_gatt_register() from main() and then return from main() (which is a valid thing to do).

jhedberg replied:
I think the only reasonable thing is to require apps to actually create their own static bt_gatt_svc variables.

jhedberg replied:
And since we're on the topic, the name should be bt_gatt_service to match the function name for registration, and the function name should probably be bt_gatt_service_register() to be consistent with Zephyr API naming conventions (think e.g. of bt_le_adv_start() rather than bt_le_start_adv())

Vudentz committed:
Bluetooth: Shell: Add gatt-show-db command

gatt-show-db shows the current set of attributes available in the GATT
database:

bt> gatt-show-db
attr 0x0011ce80 handle 0x0001 uuid 2800 perm 0x01
attr 0x0011ce94 handle 0x0002 uuid 2803 perm 0x01
attr 0x0011cea8 handle 0x0003 uuid 2a00 perm 0x01
attr 0x0011cebc handle 0x0004 uuid 2803 perm 0x01
attr 0x0011ced0 handle 0x0005 uuid 2a01 perm 0x01
attr 0x0011cde0 handle 0x0006 uuid 2800 perm 0x01
attr 0x0011cdf4 handle 0x0007 uuid 2803 perm 0x01
attr 0x0011ce08 handle 0x0008 uuid 2a05 perm 0x00
attr 0x0011ce1c handle 0x0009 uuid 2902 perm 0x03
attr 0x0011c9a0 handle 0x000a uuid 2800 perm 0x01
attr 0x0011c9b4 handle 0x000b uuid 2803 perm 0x01

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>
